File: Algorithms/239.SlidingWindowMaximum.py
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        res, q = [], []
        for i, num in enumerate(nums):
            if q and q[0] == i-k: 
                q.pop(0)
            while q and nums[q[-1]] < num:
                q.pop(-1)
            q.append(i)
            if i >= k-1:
                res.append(nums[q[0]])
        return res

File: Algorithms/239.SlidingWindowMaximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        if (k == 0) {
            return new int[0];
        }
        int[] result = new int[nums.length-k+1];
        LinkedList<Integer> q = new LinkedList<>();
        for (int i=0; i < nums.length; i++) {
            if (q.size() != 0 && q.peekFirst() == i-k) {
                q.pollFirst();
            }
            while (q.size() != 0 && nums[q.peekLast()] < nums[i]) {
                q.pollLast();
            }
            q.addLast(i);
            if (i>=k-1) {
              result[i-(k-1)] = nums[q.peekFirst()];
            }
        }
        return result;
    }
}
